cmd/plume/prerelease.go

```diff
@@ -0,0 +1,233 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"compress/bzip2"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Microsoft/azure-vhd-utils-for-go/vhdcore/validator"
+	"github.com/spf13/cobra"
+	"golang.org/x/net/context"
+
+	"github.com/coreos/mantle/auth"
+	"github.com/coreos/mantle/platform/api/azure"
+	"github.com/coreos/mantle/sdk"
+	"github.com/coreos/mantle/storage"
+)
+
+var (
+	preReleaseDryRun bool
+	cmdPreRelease    = &cobra.Command{
+		Use:   "pre-release [options]",
+		Short: "Run pre-release steps for CoreOS",
+		Long:  "Runs pre-release steps for CoreOS, such as image uploading and OS image creation, and replication across regions.",
+		RunE:  runPreRelease,
+	}
+
+	azureOpts = azure.Options{}
+)
+
+func init() {
+	cmdPreRelease.Flags().StringVar(&azureOpts.PublishSettingsFile, "publish-settings", "", "publish settings file")
+	cmdPreRelease.Flags().BoolVarP(&preReleaseDryRun, "dry-run", "n", false,
+		"perform a trial run, do not make changes")
+
+	AddSpecFlags(cmdPreRelease.Flags())
+	root.AddCommand(cmdPreRelease)
+}
+
+func runPreRelease(cmd *cobra.Command, args []string) error {
+	if len(args) > 0 {
+		return errors.New("no args accepted")
+	}
+
+	spec := ChannelSpec()
+	ctx := context.Background()
+	client, err := auth.GoogleClient()
+	if err != nil {
+		return err
+	}
+
+	src, err := storage.NewBucket(client, spec.SourceURL())
+	if err != nil {
+		return err
+	}
+
+	// Sanity check!
+	if vertxt := src.Object(src.Prefix() + "version.txt"); vertxt == nil {
+		verurl := src.URL().String() + "version.txt"
+		plog.Fatalf("File not found: %s", verurl)
+	}
+
+	plog.Printf("Running Azure pre-release...")
+
+	if err := azurePreRelease(ctx, client, src, &spec); err != nil {
+		return err
+	}
+
+	plog.Printf("Pre-release complete, run `plume release` to finish.")
+
+	return nil
+}
+
+// bunzip2 unzips file in into file out
+func bunzip2(in, out string) error {
+	inf, err := os.Open(in)
+	if err != nil {
+		return err
+	}
+
+	defer inf.Close()
+
+	bzr := bzip2.NewReader(inf)
+
+	outf, err := os.OpenFile(out, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+
+	defer outf.Close()
+
+	_, err = io.Copy(outf, bzr)
+	return err
+}
+
+// azurePreRelease runs everything necessary to prepare a CoreOS release for Azure.
+//
+// This includes uploading the vhd image to Azure storage, creating an OS image from it,
+// and replicating that OS image.
+func azurePreRelease(ctx context.Context, client *http.Client, src *storage.Bucket, spec *channelSpec) error {
```

Comment:
This function is way to big, break up the major steps into smaller functions please. Basically any time you have `if ... { log skip } else { lots of stuff }` can turn into `func foo() { if ... { log skip; return}; lots of stuff }`